kratos/processes/compute_nodal_gradient_process.cpp

```diff
@@ -26,42 +23,75 @@
+    const std::size_t local_space_dimension = r_first_element_geometry.LocalSpaceDimension();
+    const std::size_t number_of_nodes = r_first_element_geometry.PointsNumber();
+    
+    // The integration points
+    const auto& integration_method = r_first_element_geometry.GetDefaultIntegrationMethod();
+    const auto& integration_points = r_first_element_geometry.IntegrationPoints(integration_method);
+    const std::size_t number_of_integration_points = integration_points.size();
+    
+    Matrix DN_DX = ZeroMatrix(number_of_nodes, dimension);
+    Vector N = ZeroVector(number_of_nodes);
+    Matrix J0 = ZeroMatrix(dimension, local_space_dimension);
+    
+    #pragma omp parallel for firstprivate(DN_DX,  N, J0)
+    for(int i_elem=0; i_elem<static_cast<int>(mrModelPart.Elements().size()); ++i_elem) {
+        auto it_elem = it_element_begin + i_elem;
+        auto& r_geometry = it_elem->GetGeometry();
+    
+        Vector values(number_of_nodes);
+        if (mrOriginVariableDoubleList.size() > 0) {
+            for(std::size_t i_node=0; i_node<number_of_nodes; ++i_node)
+                values[i_node] = r_geometry[i_node].FastGetSolutionStepValue(mrOriginVariableDoubleList[0]);
+        } else {
+            for(std::size_t i_node=0; i_node<number_of_nodes; ++i_node)
+                values[i_node] = r_geometry[i_node].FastGetSolutionStepValue(mrOriginVariableComponentsList[0]);
+        }
         
-        const array_1d<double,TDim> grad = prod(trans(DN_DX), values);
+        // The containers of the shape functions and the local gradients
+        const auto& rNcontainer = r_geometry.ShapeFunctionsValues(integration_method);
+        const auto& rDN_DeContainer = r_geometry.ShapeFunctionsLocalGradients(integration_method);
         
-        for(std::size_t i=0; i<TDim+1; ++i) {
-            for(std::size_t k=0; k<TDim; ++k) {
-                double& val = GetGradient(geom, i,k);
+        for ( IndexType point_number = 0; point_number < number_of_integration_points; ++point_number ) {
+            // Getting the shape functions
+            noalias(N) = row(rNcontainer, point_number);
+            
+            // Getting the jacobians and local gradients
+            GeometryUtils::JacobianOnInitialConfiguration(r_geometry, integration_points[point_number], J0);
+            double detJ0;
+            Matrix InvJ0;
+            MathUtils<double>::InvertMatrix(J0, InvJ0, detJ0);
+            const Matrix& rDN_De = rDN_DeContainer[point_number];
+            GeometryUtils::ShapeFunctionsGradients(rDN_De, InvJ0, DN_DX);
+            
+            const Vector grad = prod(trans(DN_DX), values);
+            const double gauss_point_volume = integration_points[point_number].Weight() * detJ0;
+            
+            for(std::size_t i_node=0; i_node<number_of_nodes; ++i_node) {
+                for(std::size_t k=0; k<dimension; ++k) {
+                    double& val = GetGradient(r_geometry, i_node,k);
```

Comment:
this algorithm looks familiar, didn't you recently implement sth similar?

Reply:
i  think i commented this algorithm to Daniel Diez in his PR on distance calculation

Reply:
i  think i commented this algorithm to Daniel Diez in his PR on distance calculation

HOWEVER on line 84 you access many times to the database. that's somewhat slow...and you coukd at least avoid doing it once per component